docs/rfcs/blocks-in-widget-areas.md

```diff
@@ -0,0 +1,535 @@
+
+### Fetching all widget areas
+
+Fetches all registered widget areas.
+
+#### Request
+
+- Method: `GET`
+- Route: `/wp/v2/widget-areas`
+
+```
+GET /wp/v2/widget-areas
+```
+
+#### Response
+
+- Status: `200` if successful
+- Body: A list of widget area resources
+
+```json
+[
+	{
+		"id": "footer",
+		"name": "Footer",
+		"description": "",
+		"content": {
+			"raw": "<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->",
+			"rendered": "<p>Hello there!</p>"
+		},
+	},
+	{
+		"id": "sidebar",
+		"name": "Sidebar",
+		"description": "",
+		"content": {
+			"raw": "<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->",
+			"rendered": "<p>Hello there!</p>"
+		},
+	}
+]
+```
+
+### Fetching a single widget area
+
+Fetches a single registered widget area by its ID.
+
+#### Request
+
+- Method: `GET`
+- Route: `/wp/v2/widget-areas/:id`
+- Parameters:
+  - `id`: The unique identifier of the widget area to fetch
+
+```
+GET /wp/v2/widget-areas/footer
+```
+
+#### Response
+
+- Status:
+  - `200` if successful
+  - `404` if requested widget area does not exist
+- Body: A single widget area resource
+
+```json
+{
+	"id": "footer",
+	"name": "Footer",
+	"description": "",
+	"content": {
+		"raw": "<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->",
+		"rendered": "<p>Hello there!</p>"
+	},
+}
+```
+
+### Updating a single widget area
+
+Updates the blocks in a single registered widget area.
+
+#### Request
+
+- Method: `PUT`
+- Route: `/wp/v2/widget-areas/:id`
+- Parameters:
+  - `id`: The unique identifier of the widget area to update
+- Body: A single widget area resource
+
+```
+PUT /wp/v2/widget-areas/footer
+
+{ "content": { "raw": "<!-- wp:block {\"ref\":123} /-->" } }
+```
+
+#### Response
+
+- Status:
+  - `200` if successful
+  - `404` if specified widget area does not exist
+  - `400` if request body is not a valid widget area resource
+- Body: The updated widget area resource
+
+```json5
+{
+	"id": "footer",
+	"name": "Footer",
+	"description": "",
+	"content": {
+		"raw": "<!-- wp:block {\"ref\":123} /-->",
+		"rendered": "<p>This is a reusable paragraph block</p>"
+	},
+}
+```
+
+## Storage
+
+WordPress currently stores which widgets belong to which widget areas using the `'sidebars_widgets'` site option. This is a serialized PHP array that maps widget area identifiers to a list of widget identifiers that are in that area.
+
+```php
+array(
+	'footer' => array(
+		'search-2',
+		'recent-posts-2',
+		'recent-comments-2',
+	),
+	'sidebar' => array(
+		'recent-posts-2',
+		'recent-comments-2',
+	),
+	'array_version' => 3,
+)
+```
+
+When a widget area stored this way is accessed via the REST API, widgets will be represented using `core/legacy-widget` blocks. Each `core/legacy-widget` block stores a reference to the existing widget (`identifier`), and, if it is a multi widget, the attributes that that instance has (`instance`).
```

Comment:
I guess the alternative here is to have a migration script to avoid touching this site option at all. The cons for a migration script is that if you disable "Gutenberg" or disable the new widgets screen, you won't be able to fallback to the existing widgets screen, you'll lose your areas.

Also, if we migrate the storage, we'll have to rewrite all the php functions dealing with widgets.

Reply:
Based on this, I think a soft-fallback like you're proposing here is probably the best approach (it kind of maps well with the approach used by the "classic block" in the post_content editor)

Reply:
A backend based script is the way we should go here. All data would be copied to posts format and not deleted. Copy data shouldn't break anything, until you turn of the new widget screen later.

Reply:
What are the benefits of such an approach?

Reply:
The backend / rest api is the data layer in this application and should deal with saving data and migrating it. Javascript is the visual layer here and shouldn't have to deal with a complex migration. Also, if you have get the old data, you would have to get that from the api anyway. I think the migration should happen in PHP and javascript should only ever deal with migrated widgets. 

There is already a long history of migrations like this happening in the PHP, the team split is a good example. Any migration is going to need to give a action / filter so that PHP develops are right some custom bahavior on the migration if they so wish?

Reply:
> The backend / rest api is the data layer in this application and should deal with saving data and migrating it. Javascript is the visual layer here and shouldn't have to deal with a complex migration. Also, if you have get the old data, you would have to get that from the api anyway. I think the migration should happen in PHP and javascript should only ever deal with migrated widgets.

I think we all agree with this, my preference right now is to avoid a migration script though and have this behavior tied to the REST API instead. It seems simple enough and it leaves the door open for downgrades/classic widget.

Reply:
@youknowriad I am confused how that api would work then. Let's have a widgets api. By default, it loads data on options. When you click save, it copies that data to the post table returns from the post table going forward. The migration in effect happens on save. But then you could have only parts loading from posts and others from options. That would be a mess and generates tonnes of edges cases. 

You know not want a migration, but to cleaner and safer for the user if it happens that way. It cleaner for gutenberg team too, as they only ever deal with the migrated (copied) widgets that  are already blocks and not the current ones.

Reply:
> But then you could have only parts loading from posts and others from options. That would be a mess and generates tonnes of edges cases.

Yes, that's not great and that's not what is proposed here.

The idea is that the API would fetch the post if the area references an id (post save) or runs the migration script if the widgets are still on the site option table.

Reply:
@youknowriad So instead of running the migration on upgrade, you want to run in on demand when you access the api? What if users never access the api? They never go onto the edit screen and the data is never migrated? We could be a situation where in 10 years time, we still have data in the option table that was never migrated. Data constancy is important. What do you say to developers who want to access / use that data, check in both places, that is a nightmare to work with. 

Other wordpress upgrade have migrated data, like the term split. See this for example - https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/upgrade.php#L685-L824

Also, a maintainer of multisite, I am going to say that large hosts with multisites are going to want this migration to happen in much more controlled way.

Reply:
> 
> Yes, that's not great and that's not what is proposed here.
> 

This is a request for comments, which is a proposal, not a fixed plan.

Reply:
I discussed this matter in detail with other members of the core team at wordcamp London, they agreed that a migration like this should happen in php. It is simply not acceptable to have the data not loading from two places. If I wanted to filter the value of widget saving in the database, how would I do that if it was stored in two places? How can I trust where is the data loading from? How can a query that data in the post table if I can’t trust it is there?

There needs to be a migration and then a flag / option set. Developers can use this flag to know that data has moved. 

Also, like term splitting, that an action in php will be required for developers (like jet pack and widget visabily), so they can transform / modify data. For example, what if you want to migrate a custom widget to a custom block. We should be pushing developers to do this!

Reply:
If the migration happens as soon as the REST API is used, and if the new widgets.php screen is using the the REST API, then all sidebars, and all widgets within, will be migrated in one hit the first time a user visits the widgets.php screen. Is that right?

I develop a plugin that registers its own sidebar to use as a "holding area" for widgets that are displayed within a menu (when displaying the menu, widgets are extracted from the sidebar and organised into multiple sub menus). Unlike a normal sidebar where all of the widgets are output in one group using dynamic_sidebar(), we "pluck" widgets from our "holding area" and display them individually.

If I understand correctly, then once the migration has happened, all widgets (previously represented in the database as their own entities) will be grouped together into a single post. I think this will make it difficult to single out an individual widget instance and display it on its own (the same goes for saving a widget). That is, unless the REST API will be given an end point for displaying/saving a **single** widget instance?

I'm concerned once the automatic migration happens, I'll be left in a situation which makes it difficult to recover from, due to the differing ways that widgets are stored currently (almost individually) to how is proposed (grouped together). For backwards compatibility, could we have a "is_legacy_sidebar" parameter for "register_sidebar" which would prevent the sidebar from automatically being migrated? It may also prevent blocks from being added to the sidebar. That would give me, and anyone in a similar situation some breathing space to update our plugins once things are more ironed out. The ultimate goal would be to allow blocks to be added to a menu in the same way as widgets are.

Reply:
As state right now, the migration only happens when you "save"/"update" the widget area.  but the idea shared by @spacedmonkey asking for a migration script that runs on "upgrade" is growing on me. This would only work if there's a "downgrade" script as well which should be called:

 - If we disable the Gutenberg plugin (during the plugin phase before it hits Core)
 - If we use a plugin similar to "classic editor" for widgets.

What do you think @noisysocks (when you're back from AFK :) )

Reply:
I can speak from personal experience that there will be sites with hundreds potentially thousands of widgets. I would insist that a WP CLI based option to upgrade to the new widget storage would be needed.

If that's the case, a PHP migration route would be needed, so why not run it on WP upgrade?

Reply:
Good discussion here! Thanks for hashing out the details.

My understanding is that @spacedmonkey suggests that instead of _lazily_ migrating individual items in `sidebars_widgets` when a widget area is saved, we ought to _eagerly_ migrate all items in `sidebars_widgets` when WordPress is installed or when Gutenberg is activated. This means that we don't end up in a scenario where some widget areas are stored different to other widget areas.

I think this makes sense and I'll update the RFC accordingly. 

> If I understand correctly, then once the migration has happened, all widgets (previously represented in the database as their own entities) will be grouped together into a single post.

Widgets are not currently stored in the database as their own entities—they're serialised into a PHP array which is stored as a site option. Accessing or displaying a single widget is fairly difficult right now.

> For backwards compatibility, could we have a "is_legacy_sidebar" parameter for "register_sidebar()" which would prevent the sidebar from automatically being migrated? 

@megamenu: Could you elaborate on why you would like this? What's proposed is that existing widgets will work as-is after a WordPress upgrade via the Legacy Widget block.

> I think no matter the solution we came up with, it is necessary to be able to go back to the old screen without loosing content.

Definitely, but this is separate to the discussion about whether we migrate `sidebars_widgets` lazily or eagerly. Either way we will need a way to "un-migrate" the data. See discussion in https://github.com/WordPress/gutenberg/pull/14812#discussion_r272088017.

Reply:
This RFC keeps mentioning the widgets screen and ignores the customiser. We have been pushing developers to use the customiser for years. Any replacement of widgets, should focus on pushing them to the customiser and hide the old widget screen. I believe there was even a core ticket from @westonruter to hide the old widget screen. 

New installs of WordPress should never see the old widget screen either. We need a feature flag in option name, use_legacy_widgets. This will the flag to show the new block areas or not. 

This plugin for legacy widgets should filter this value. It should another plugin apart from classic editor. You might want to opt out of widget but keep the new editor.

Reply:
This RFC focuses on the underlying infrastructure which both screens will use. Right now the assumption is that the widget screen is [here to stay](https://make.wordpress.org/core/2018/12/08/9-priorities-for-2019/).

Reply:
> Widgets are not currently stored in the database as their own entities—they're serialised into a PHP array which is stored as a site option. Accessing or displaying a single widget is fairly difficult right now.

They're not stored entirely separately, but it is possible to access/display/save a single widget _fairly_ easily. These are the functions (this one and the next 5 or so) that I am concerned about not being able to make compatible with the proposed changes:

https://plugins.trac.wordpress.org/browser/megamenu/trunk/classes/widget-manager.class.php?rev=1317340#L512

My plugin allows users to add widgets to their menu. We store all of these widgets within the same sidebar, and pick and choose them individually to output in the correct place. We also provide our own interface for adding and editing these widgets (there is no need for the user to be able to edit those widgets on the widgets.php screen).

The "StoreFront Mega Menus" (an official WooCommerce plugin) also stores/displays widgets in a similar way, although they have completely hidden their own sidebar from the widgets.php screen.

> [r.e. is_legacy_sidebar suggestion) Could you elaborate on why you would like this? What's proposed is that existing widgets will work as-is after a WordPress upgrade via the Legacy Widget block.

I suggested this because if there will be built in support for "legacy" widgets anyway, I will not have to ask my users to install the "Legacy Widget" plugin - I can simply mark the sidebar as "legacy" (and even hide it from the widgets.php screen).